### PR TITLE
Skip softmax and topk

### PR DESCRIPTION
**Pull request by bricksdont**
Skip softmax by default for greedy decoding, and adding option to skip topk:
- this change only affects inference
- only affects greedy decoding

Skipping softmax is significantly faster, skipping topk is significantly slower, but it could be faster if the MXnet implementation of argmin changes.

#### Pull Request Checklist ##
- [x] Changes are complete (if posting work-in-progress code, prefix your pull request title with '[WIP]'
until you can check this box.
- [x] Unit tests pass (`pytest`)
- [x] Were system tests modified? If so did you run these at least 5 times to account for the variation across runs?
- [x] System tests pass (`pytest test/system`)
- [x] Passed code style checking (`./style-check.sh`)
- [x] You have considered writing a test
- [x] Updated major/minor version in `sockeye/__init__.py`. Major version bump if this is a backwards incompatible change.
- [x] Updated CHANGELOG.md

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: sockeye/utils.py [hunk 1]
```diff
@@ -251,6 +251,28 @@ def variance(self) -> float:
             return self._M2 / self._count
 
 
+def top1(scores: mx.nd.NDArray,
+         offset: mx.nd.NDArray) -> Tuple[mx.nd.NDArray, mx.nd.NDArray, mx.nd.NDArray]:
+    """
+    Get the single lowest element per sentence from a `scores` matrix. Expects that
+    beam size is 1, for greedy decoding.
+
+    NOTE(mathmu): The current implementation of argmin in MXNet much slower than topk with k=1.
```
Comment (fhieber) [inline sockeye/utils.py]: In that case I think we should use topk then and add a TODO to change to argmin once its sped up.

File: test/integration/test_seq_copy_int.py [hunk 2]
```diff
@@ -33,6 +33,22 @@
      " --decode-and-evaluate 0",
      "--beam-size 2 --softmax-temperature 0.01",
      True, False, False),
+    # "Vanilla" LSTM encoder-decoder with attention, greedy and skip softmax
+    ("--encoder rnn --decoder rnn --num-layers 1 --rnn-cell-type lstm --rnn-num-hidden 8 --num-embed 4 "
+     " --rnn-attention-type mlp"
+     " --rnn-attention-num-hidden 8 --batch-size 2 --loss cross-entropy --optimized-metric perplexity --max-updates 2"
+     " --checkpoint-frequency 2 --optimizer adam --initial-learning-rate 0.01 --batch-type sentence "
+     " --decode-and-evaluate 0",
+     "--beam-size 2 --softmax-temperature 0.01",
```
Comment (fhieber) [inline test/integration/test_seq_copy_int.py]: I don't see the `--skip-softmax` here; also, using --softmax-temperature in this test is quite confusing then :)

File: sockeye/inference.py [hunk 3]
```diff
@@ -1377,10 +1412,13 @@ def _combine_predictions(self,
 
         # combine model predictions and convert to neg log probs
         if len(self.models) == 1:
-            neg_logprobs = -mx.nd.log(probs[0])  # pylint: disable=invalid-unary-operand-type
+            if self.beam_size == 1 and self.skip_softmax:
+                neg_probs = -probs[0]
+            else:
+                neg_probs = -mx.nd.log(probs[0])  # pylint: disable=invalid-unary-operand-type
```
Comment (fhieber) [inline sockeye/inference.py]: these are still negative log probs, no?

File: test/integration/test_seq_copy_int.py (hunk 2 above)
Comment (bricksdont) [reply test/integration/test_seq_copy_int.py]: `--skip-softmax` is not an option anymore, but the default behaviour. Sorry about the confusion; it is something I changed with the third commit.

File: sockeye/inference.py (hunk 3 above)
Comment (bricksdont) [reply sockeye/inference.py]: If I understand correctly, if log is not applied they will not be in log space.

File: sockeye/utils.py (hunk 1 above)
Comment (bricksdont) [reply sockeye/utils.py]: The current default behaviour is to use topk for beam size 1, using top1 is a CLI option: `--skip-topk`.

File: sockeye/inference.py (hunk 3 above)
Comment (fhieber) [reply sockeye/inference.py]: sure, so neither name can be entirely correct :)

File: sockeye/inference.py [hunk 4]
```diff
@@ -79,6 +83,8 @@ def __init__(self,
         self.max_input_length, self.get_max_output_length = models_max_input_output_length([self],
                                                                                            max_output_length_num_stds,
                                                                                            forced_max_output_len=forced_max_output_len)
+        self.skip_softmax = skip_softmax
```
Comment (tdomhan) [inline sockeye/inference.py]: you should probably add a `check_condition` on the `beam_size`

File: sockeye/inference.py [hunk 5]
```diff
@@ -394,6 +406,13 @@ def load_models(context: mx.context.Context,
     if checkpoints is None:
         checkpoints = [None] * len(model_folders)
 
+    # skip softmax for a single model,
+    if len(model_folders) == 1:
+        skip_softmax = True
```
Comment (tdomhan) [inline sockeye/inference.py]: probably makes sense to only do this for `beam_size == 1` explicitly

File: sockeye/inference.py (hunk 3 above)
Comment (bricksdont) [reply sockeye/inference.py]: Oh I see - yes you are right. Want me to change it?

File: sockeye/inference.py [hunk 6]
```diff
@@ -67,7 +69,9 @@ def __init__(self,
                  max_output_length_num_stds: int = C.DEFAULT_NUM_STD_MAX_OUTPUT_LENGTH,
                  decoder_return_logit_inputs: bool = False,
                  cache_output_layer_w_b: bool = False,
-                 forced_max_output_len: Optional[int] = None) -> None:
+                 forced_max_output_len: Optional[int] = None,
+                 skip_softmax: bool = False,
+                 skip_topk: bool = False) -> None:
```
Comment (tdomhan) [inline sockeye/inference.py]: `skip_topk` is only used in Translator but not in InfereceModel. Therefore it should not be a member of InferenceModel, but rather be directly passed to Translator.

File: test/integration/test_seq_copy_int.py (hunk 2 above)
Comment (tdomhan) [reply test/integration/test_seq_copy_int.py]: you should update the comment then. Also: It's `--beam-size 2` and therefore not greedy ;)

File: sockeye/inference.py [hunk 7]
```diff
@@ -67,18 +68,22 @@ def __init__(self,
...
+                 forced_max_output_len: Optional[int] = None,
+                 skip_softmax: bool = False) -> None:
         super().__init__(config)
         self.params_fname = params_fname
         self.context = context
         self.beam_size = beam_size
         utils.check_condition(beam_size < self.config.vocab_target_size,
                               'The beam size must be smaller than the target vocabulary size.')
+        if skip_softmax:
+            utils.check_condition(beam_size == 1, 'Skipping softmax does not have any effect for beam size > 1')
```
Comment (tdomhan) [inline sockeye/inference.py]: as this will raise a RuntimeError, maybe: "Softmax skipping can only be enabled with beam size of 1."

File: sockeye/inference.py [hunk 8]
```diff
@@ -394,6 +403,13 @@ def load_models(context: mx.context.Context,
     if checkpoints is None:
         checkpoints = [None] * len(model_folders)
 
+    # skip softmax for a single model,
+    if len(model_folders) == 1 and beam_size == 1:
+        skip_softmax = True
+    else:
+        # but not for an ensemble
```
Comment (tdomhan) [inline sockeye/inference.py]: you should probably warn the user with a logging message in this case.

File: sockeye/inference.py (hunk 7 above)
Comment (tdomhan) [reply sockeye/inference.py]: also, actually now skip_softmax is no longer a user facing parameter. So I in that case I'd say make it an assertion.

File: sockeye/inference.py [hunk 9]
```diff
@@ -981,7 +1001,8 @@ def __init__(self,
                  restrict_lexicon: Optional[lexicon.TopKLexicon] = None,
                  avoid_list: Optional[str] = None,
                  store_beam: bool = False,
-                 strip_unknown_words: bool = False) -> None:
+                 strip_unknown_words: bool = False,
+                 skip_topk: bool = False,) -> None:
```
Comment (tdomhan) [inline sockeye/inference.py]: you probably want to remove that comma

File: sockeye/inference.py [hunk 10]
```diff
@@ -1005,6 +1026,9 @@ def __init__(self,
         self.interpolation_func = self._get_interpolation_func(ensemble_mode)
         self.beam_size = self.models[0].beam_size
         self.batch_size = self.models[0].batch_size
+        # skip softmax for a single model, but not for an ensemble
+        self.skip_softmax = True if len(models) == 1 else False
```
Comment (tdomhan) [inline sockeye/inference.py]: let's decide on `skip_softmax` in just a single location, probably in `load_models` and reuse it after.

File: sockeye/inference.py [hunk 11]
```diff
@@ -1352,9 +1380,14 @@ def _decode_step(self,
             # Compute logits and softmax with restricted vocabulary
             if self.restrict_lexicon:
                 logits = model.output_layer(decoder_outputs, out_w, out_b)
-                probs = mx.nd.softmax(logits)
+                if self.beam_size == 1 and self.skip_softmax:
```
Comment (tdomhan) [inline sockeye/inference.py]: wouldn't it be enough to use `skip_softmax`, as it should only be true with beam_size 1?

File: sockeye/inference.py [hunk 12]
```diff
@@ -1005,6 +1028,9 @@ def __init__(self,
         self.interpolation_func = self._get_interpolation_func(ensemble_mode)
         self.beam_size = self.models[0].beam_size
         self.batch_size = self.models[0].batch_size
+        # skip softmax for a single model, but not for an ensemble
+        self.skip_softmax = self.models[0].skip_softmax
+        self.skip_topk = skip_topk
```
Comment (tdomhan) [inline sockeye/inference.py]: maybe a final check_condition to cover the case of someone creating a `Translator` without `load_models`.

```
if skip_softmax:
  check_condition(len(self.models) == 0 and self.beam_size ==1, "...")
```

File: sockeye/inference.py [hunk 13]
```diff
@@ -1377,10 +1412,13 @@ def _combine_predictions(self,
 
         # combine model predictions and convert to neg log probs
         if len(self.models) == 1:
-            neg_logprobs = -mx.nd.log(probs[0])  # pylint: disable=invalid-unary-operand-type
+            if self.beam_size == 1 and self.skip_softmax:
```
Comment (tdomhan) [inline sockeye/inference.py]: is this check necessary?

File: sockeye/inference.py [hunk 14]
```diff
@@ -394,6 +403,15 @@ def load_models(context: mx.context.Context,
...
         checkpoints = [None] * len(model_folders)
 
+    # skip softmax for a single model,
+    if len(model_folders) == 1 and beam_size == 1:
+        skip_softmax = True
+        logger.info("Enabled skipping softmax for a single model and greedy decoding.")
+    else:
+        # but not for an ensemble or beam search
+        skip_softmax = False
+        logger.info("Disabled skipping softmax for several models or beam size larger than 1.")
```
Comment (fhieber) [inline sockeye/inference.py]: I think this warning can be confusing if a user does NOT set skip_softmax and runs with the default settings (beam_size=5). Maybe only print this warning when skip_softmax==True? Or remove it entirely

File: sockeye/utils.py (hunk 1 above)
Comment (fhieber) [reply sockeye/utils.py]: but aren't you saying that this is slower than using mx.nd.topk? What's the point of this option if its slower?
